dvc/repo/metrics/show.py

```diff
@@ -80,71 +129,23 @@ def show(
     for branch in self.scm.brancher(
         all_branches=all_branches, all_tags=all_tags
     ):
-        astages = self.active_stages()
-        outs = [out for stage in astages for out in stage.outs]
-
-        if path:
-            outs = find_output_by_path(
-                self, path, outs=outs, recursive=recursive
-            )
-
-        metrics = filter(lambda o: o.metric, outs)
-        stages = set()
-        entries = []
-
-        for o in metrics:
-            if not typ and isinstance(o.metric, dict):
-                t = o.metric.get(o.PARAM_METRIC_TYPE, typ)
-                x = o.metric.get(o.PARAM_METRIC_XPATH, xpath)
-            else:
-                t = typ
-                x = xpath
-            entries.append((o.path, t, x))
-            stages.add(o.stage.path)
+        entries = _collect_metrics(self, path, recursive, typ, xpath)
 
         if path and not entries:
             if os.path.isdir(path):
                 logger.warning(
                     "Path '{path}' is a directory. "
                     "Consider running with '-R'.".format(path=path)
                 )
-                return {}
-
-        for fname, t, x in entries:
-            if stages:
-                for stage in stages:
-                    self.checkout(stage, force=True)
-
-            rel = os.path.relpath(fname)
-            metric = _read_metric(fname, typ=t, xpath=x)
-            if not metric:
                 continue
 
-            if branch not in res:
-                res[branch] = {}
-
-            res[branch][rel] = metric
-
-    for branch, val in res.items():
-        if all_branches or all_tags:
-            logger.info("{}:".format(branch))
-        for fname, metric in val.items():
-            logger.info("\t{}: {}".format(fname, metric))
+        metrics = _read_metrics(self, entries)
+        if metrics:
+            res[branch] = metrics
 
-    if res:
-        return res
-
-    if path and os.path.isdir(path):
-        return res
+    if not res:
+        if path:
+            raise BadMetricFileError(path)
```

Comment:
this exception is misleading if path is a directory, I would double check that we don't break something for @tdeboissiere by removing like 137-138